keps/0031-namespace-management.md

```diff
@@ -0,0 +1,167 @@
+
+### Helm
+
+Helm v1 and v2 provided an auto-provision capability.  The initial release of Helm v3 removed this capability and now requires users to create namespaces prior to installing a helm chart. Much detail is written in comments for [helm issue 6794](https://github.com/helm/helm/issues/6794).  Helm v3 [reintroduced the ability](https://github.com/helm/helm/issues/6794) to create a namespace through the use of `--create-namespace` flag.
+
+### The Many Meanings of a Namespace
+
+Challenages to auto-provision of namespaces are complicated by the many reasons / values for creating them which includes:
+
+1. Logical separation of work (separating 1000s of pods into 10s of buckets for management)
+1. Quota management
+1. Security separation
+1. Metadata (where labels and annotations provide meaning and value to controllers managing resources in a namespace)
+
+Based on the last 3 reasons, it is common for an administrator to create namespaces to manage the other concerns associated with a namespace.  It would be challeging to manage a cluster where dozen or more namespaces are adhoc being created (or deleted) which may require a host of other configurations around quota and security.
+
+### Hierarchical Namespace Controller
+
+Currently in incubation, Kubernetes is introducing a Hierarchical Namespace Controller to more easily create and manage namespaces in the cluster, even if you don't have cluster-level permission to create namespaces, which is something to keep in mind in the future.  See https://github.com/kubernetes-sigs/multi-tenancy/tree/master/incubator/hnc
+
+## Proposals
+
+Prior to making KEP `provisional` we will chose a namespace creation proposal to move forward with and move the 2nd proposal to alternative.  For now, we have 2 competing namespace creation proposals.  
+In addition there is a proposal for the use of multiple namespaces.
+
+### Proposal: Single Namespace Support
+
+There is a desire to support multiple namespaces, however it requires the addition of cluster-wide resource support in KUDO.  Based on this and the desire to not break Kubernetes garbase collection, KUDO currenly only supports 1 namespace.  Which results in the following explicit rules for operators until a cluster-wide support is provided:
+
+* All manifests with metadata defining a namespace will be ignored or overridden.  The namespace that is being used by KUDO to install the operator is the only namespace that resources will be installed to.
+* All namespace manifests (manifest of `"kind": "Namespace"`) will have no affect as they can not be used by the operator.
+* All namespace manifests will be flagged by `package verify` as a warning
```

Comment:
Shouldn't they be flagged as an error? I'd rather have them fail loud and clear.

For the installation, i agree that "have no affect" may be ok, as if we would fail we might break existing operators.

Reply:
I debated this as well... I'm good with a failure and will update